toolkit/recordreplay/Network.cpp

```diff
@@ -102,17 +105,336 @@ already_AddRefed<nsIStreamListener> WrapNetworkStreamListener(nsIStreamListener*
   return tee.forget();
 }
 
+/**
+ * The gecko-standard nsInputStreamTee implementation you get from 'NS_NewInputStreamTee'
+ * only implements 'nsIInputStream', but for things to behave well when wrapping network
+ * request bodies, we need to implement 'nsIAsyncInputStream' and 'nsIInputStreamLength'
+ * at a minimum.
+ *
+ * NOTE: Much of the logic here is copied and tweaked from other stream implementations.
+ */
+class ReplayInputStreamTee final : public nsIAsyncInputStream,
+                                   public nsIInputStreamLength,
+                                   public nsIInputStreamCallback,
+                                   public nsIAsyncInputStreamLength,
+                                   public nsIInputStreamLengthCallback {
+ public:
+  ReplayInputStreamTee(nsIInputStream* aStream, nsIOutputStream* aSink);
+
+  NS_DECL_THREADSAFE_ISUPPORTS
+  NS_DECL_NSIINPUTSTREAM
+  NS_DECL_NSIASYNCINPUTSTREAM
+  NS_DECL_NSIINPUTSTREAMLENGTH
+  NS_DECL_NSIINPUTSTREAMCALLBACK
+  NS_DECL_NSIASYNCINPUTSTREAMLENGTH
+  NS_DECL_NSIINPUTSTREAMLENGTHCALLBACK
+
+ private:
+  nsCOMPtr<nsIInputStream> mSource;
+  nsCOMPtr<nsIOutputStream> mSink;
+  ~ReplayInputStreamTee() = default;
+
+  // Optional interfaces implemented by source.
+  nsCOMPtr<nsIAsyncInputStream> mAsyncSource;
+  nsCOMPtr<nsIInputStreamLength> mSourceLength;
+  nsCOMPtr<nsIAsyncInputStreamLength> mAsyncSourceLength;
+
+  // This is used for AsyncWait and it's protected by mutex.
+  mozilla::Mutex mCallbackMutex;
+  nsCOMPtr<nsIInputStreamCallback> mAsyncWaitCallback;
+  nsCOMPtr<nsIInputStreamLengthCallback> mAsyncLengthWaitCallback;
+};
+
+NS_IMPL_ADDREF(ReplayInputStreamTee)
+NS_IMPL_RELEASE(ReplayInputStreamTee)
+NS_INTERFACE_MAP_BEGIN(ReplayInputStreamTee)
+  NS_INTERFACE_MAP_ENTRY_AMBIGUOUS(nsISupports, nsIAsyncInputStream)
```

Comment:
What does the ambiguous specifier mean here?  It's not clear this class supports `nsIAsyncInputStream`?